sdk/daml-script/test/daml/upgrades/ContractKeys.daml

```diff
@@ -3,15 +3,64 @@
 
 module ContractKeys (main) where
 
+import DA.Text (isInfixOf)
+
 import UpgradeTestLib
+
 import qualified V1.ContractKeys as V1
 import qualified V2.ContractKeys as V2
+import qualified V1.ChangedKeyExpr as V1
+import qualified V2.ChangedKeyExpr as V2
+import qualified V1.UpgradedContractKeys as V1
+import qualified V2.UpgradedContractKeys as V2
 
 {- PACKAGE
 name: contract-key-upgrades
 versions: 2
 -}
 
+main : TestTree
+main = tests
+  [ subtree "Unchanged key type"
+    [ subtree "Unchanged key expression"
+      [ ("queryContractKey, src=v1 tgt=v2", queryKeyUnchanged)
+      , ("exerciseByKeyCmd, src=v1 tgt=v2", exerciseCmdKeyUnchanged)
+      , ("fetchByKey, src=v1 tgt=v2 ", fetchKeyUnchanged)
+      , ("exerciseByKey, src=v1 tgt=v2", exerciseUpdateKeyUnchanged)
+      ]
+    , subtree "Changed key expression"
+      [ subtree "Unchanged key value"
+        [ ("queryContractId, src=v1 tgt=v2", queryKeyChangedExprSameValue)
+        , ("queryContractKey, src=v1 tgt=v2", qckKeyChangedExprSameValue)
+        , ("fetch, src=v1 tgt=v2", fetchKeyChangedExprSameValue)
+        , ("fetchByKey, src=v1 tgt=v2", fbkKeyChangedExprSameValue)
+        , ("exercise, src=v1 tgt=v2", exerciseKeyChangedExprSameValue)
+        , ("exerciseByKey, src=v1 tgt=v2", ebkKeyChangedExprSameValue)
+        , ("exerciseCmd, src=v1 tgt=v2", exerciseCmdKeyChangedExprSameValue)
+        , ("exerciseByKeyCmd, src=v1 tgt=v2", ebkCmdKeyChangedExprSameValue)
+        ]
+      , subtree "Changed key value"
+        [ broken ("queryContractId, src=v1 tgt=v2", queryKeyChangedExprChangedValue)
+        , broken ("queryContractKey, src=v1 tgt=v2", qckKeyChangedExprChangedValue)
+        , ("fetch, src=v1 tgt=v2", fetchKeyChangedExprChangedValue)
+        , broken ("fetchByKey, src=v1 tgt=v2", fbkKeyChangedExprChangedValue)
+        , ("exercise, src=v1 tgt=v2", exerciseKeyChangedExprChangedValue)
+        , broken ("exerciseByKey, src=v1 tgt=v2", ebkKeyChangedExprChangedValue)
+        , ("exerciseCmd, src=v1 tgt=v2", exerciseCmdKeyChangedExprChangedValue)
+        , broken ("exerciseByKeyCmd, src=v1 tgt=v2", ebkCmdKeyChangedExprChangedValue)
+        ]
+      ]
+    ]
+  , subtree "Changed key type"
+    [ subtree "Unchanged key value (modulo trailing `None`s)"
+      [ broken ("queryContractKey, src=v1 tgt=v2", queryKeyUpgraded)
+      , broken ("exerciseByKeyCmd, src=v1 tgt=v2", exerciseCmdKeyUpgraded)
```

Comment:
these two cases are broken on `main` even though they work on `main-2.x`

Reply:
Good to know, I think we should create explicit rows for these if they dont already exist in the spreadsheet, so we can mark this behaviour as broken publically

Reply:
Turns out, this is probably because key upgrading is not ported to 3x, unsure when/if it ever will be given non unique contract keys

Reply:
hm in that case, should we be concerned that the two cases below (`fetchKeyUpgraded` and `exerciseUpdateKeyUpgraded`) _do_ seem to work?

Reply:
Ah, possibly

Reply:
@paulbrauner-da What do you think?

Reply:
I think the commands fail because we haven't ported https://github.com/digital-asset/daml/pull/19780/files#diff-f1459a53d7980874919e46b017e19dfb5dc47c48f49494bab3422d55e62ac48b, but inside the choice bodies it works because of https://github.com/digital-asset/daml/pull/19668. 

But then I think if we allow this, we will run into the problem that https://github.com/digital-asset/daml/pull/19792 fixes in 2.x, but this is a hack and we want to solve it properly with value normalization in 3.x.

That said, as long as we forbid it at upload time, that shouldn't be reachable.